### ISSUE TITLE
Backend failure during the re-election process causes Vault cluster to become unusable.

### ISSUE THREAD

**Issue by vespian** [hashicorp/vault#4491]
**Environment:**
* 3 Vault instances backed by 3-node ZK quorum
* Vault Version: Vault v0.8.3 ('84ec52693502eef900cc531c59f0670bd3f5cff3')
* Operating System/Architecture: x86_64

**Vault Config File:**
```hcl
disable_mlock = true

backend "zookeeper" {
  address = "zk-1.zk:2181,zk-2.zk:2181,zk-3.zk:2181,zk-4.zk:2181,zk-5.zk:2181"
  advertise_addr = "https://172.16.0.1:8200"
  path = "dcos/vault/default"
  znode_owner = "digest:<some-user>:<some-pass>"
  auth_info = "digest:<some-user>:<some-pass>"
}

listener "tcp" {
  address = "127.0.0.1:8200"
  cluster_address = "172.16.0.1:8201"
  tls_disable = 1
}
```

**Startup Log Output:**
```text
Apr 27 16:16:28 master1 vault[22918]: ==> Vault server configuration:
Apr 27 16:16:28 master1 vault[22918]: Cgo: disabled
Apr 27 16:16:28 master1 vault[22918]: Cluster Address: https://172.16.0.1:8201
Apr 27 16:16:28 master1 vault[22918]: Listener 1: tcp (addr: "127.0.0.1:8200", cluster address: "172.16.0.1:8201", tls: "disabled")
Apr 27 16:16:28 master1 vault[22918]: Log Level: trace
Apr 27 16:16:28 master1 vault[22918]: Mlock: supported: true, enabled: false
Apr 27 16:16:28 master1 vault[22918]: Redirect Address: https://172.16.0.1:8200
Apr 27 16:16:28 master1 vault[22918]: Storage: zookeeper (HA available)
Apr 27 16:16:28 master1 vault[22918]: Version: Vault v0.8.3
Apr 27 16:16:28 master1 vault[22918]: Version Sha: 84ec52693502eef900cc531c59f0670bd3f5cff3
Apr 27 16:16:28 master1 vault[22918]: ==> Vault server started! Log data will stream in below:
Apr 27 16:16:28 master1 vault[22918]: 2018/04/27 16:16:28.169802 [TRACE] physical/cache: creating LRU cache: size=32768
Apr 27 16:16:28 master1 vault[22918]: 2018/04/27 16:16:28.169980 [TRACE] cluster listener addresses synthesized: cluster_addresses=[172.16.0.1:8201]
```

**Expected Behavior:**
Zookeeper/backend issues during leader reelection do not cause Vault cluster to become unreachable.

**Actual Behavior:**
Please see the logs from the moment of a failure and imediatelly after:
```text
Apr 27 16:22:00 master1 vault[22918]: 2018/04/27 16:22:00.228100 [TRACE] forwarding: successful heartbeat
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.648177 [WARN ] transport: http2Client.notifyError got notified that the client transport was broken EOF.
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.648284 [TRACE] core: creating rpc dialer: host=fw-b34dfde0-ca7c-7bd2-dfdf-9e2ce8e08425
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.649161 [WARN ] grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 172.16.0.2:8201: getsockopt: connection refused"; Reconnecting to
{172.16.0.2:8201 0  <nil>}
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.662772 [INFO ] core: acquired lock, enabling active operation
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.664012 [TRACE] core: generating cluster private key
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.674921 [TRACE] core: generating local cluster certificate
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.704861 [INFO ] core: post-unseal setup starting
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.704886 [TRACE] core: clearing forwarding clients
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.704907 [TRACE] core: done clearing forwarding clients
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.704919 [TRACE] forwarding: stopping heartbeating
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.705047 [WARN ] get error from resetTransport context canceled, transportMonitor returning
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.705066 [WARN ] grpc: addrConn.transportMonitor exits due to: context canceled
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.705354 [INFO ] core: loaded wrapping token key
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.705378 [INFO ] core: successfully setup plugin catalog: plugin-directory=
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.712728 [INFO ] core: successfully mounted backend: type=kv path=secret/
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.712957 [INFO ] core: successfully mounted backend: type=system path=sys/
Apr 27 16:22:13 master1 vault[22918]: 2018/04/27 16:22:13.712994 [INFO ] core: successfully mounted backend: type=cubbyhole path=cubbyhole/
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14 Recv loop terminated: err=EOF
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14 Send loop terminated: err=<nil>
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14.068094 [INFO ] core: pre-seal teardown starting
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14.068114 [INFO ] core: cluster listeners not running
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14.068203 [INFO ] core: pre-seal teardown complete
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14.068237 [ERROR] core: post-unseal setup failed: error=error fetching response-wrapping policy from store: failed to read policy: zk: connection closed
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14 Failed to connect to 172.16.0.2:2181: dial tcp 172.16.0.2:2181: getsockopt: connection refused
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14 Connected to 172.16.0.3:2181
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14 Authentication failed: EOF
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14 Connected to 172.16.0.1:2181
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14 Authentication failed: EOF
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14 Connected to 172.16.0.3:2181
Apr 27 16:22:14 master1 vault[22918]: 2018/04/27 16:22:14 Authentication failed: EOF
Apr 27 16:22:15 master1 vault[22918]: 2018/04/27 16:22:15 Connected to 172.16.0.1:2181
Apr 27 16:22:15 master1 vault[22918]: 2018/04/27 16:22:15 Authentication failed: EOF
Apr 27 16:22:15 master1 vault[22918]: 2018/04/27 16:22:15 Failed to connect to 172.16.0.2:2181: dial tcp 172.16.0.2:2181: getsockopt: connection refused
Apr 27 16:22:15 master1 vault[22918]: 2018/04/27 16:22:15 Connected to 172.16.0.3:2181
Apr 27 16:22:15 master1 vault[22918]: 2018/04/27 16:22:15 Authentication failed: EOF
Apr 27 16:22:15 master1 vault[22918]: 2018/04/27 16:22:15 Connected to 172.16.0.1:2181
Apr 27 16:22:15 master1 vault[22918]: 2018/04/27 16:22:15 Authentication failed: EOF
Apr 27 16:22:15 master1 vault[22918]: 2018/04/27 16:22:15 Connected to 172.16.0.3:2181
Apr 27 16:22:15 master1 vault[22918]: 2018/04/27 16:22:15 Authentication failed: EOF
Apr 27 16:22:15 master1 vault[22918]: 2018/04/27 16:22:15.154421 [ERROR] core: failed to acquire lock: error=zk: could not connect to a server
Apr 27 16:22:16 master1 vault[22918]: 2018/04/27 16:22:16 Connected to 172.16.0.1:2181
Apr 27 16:22:16 master1 vault[22918]: 2018/04/27 16:22:16 Authenticated: id=99932284807544887, timeout=4000
Apr 27 16:22:16 master1 vault[22918]: 2018/04/27 16:22:16 Re-submitting `1` credentials after reconnect
Apr 27 16:22:16 master1 vault[22918]: 2018/04/27 16:22:16.159659 [TRACE] core: found new active node information, refreshing
Apr 27 16:22:16 master1 vault[22918]: 2018/04/27 16:22:16.160189 [TRACE] core: parsing information for new active node: active_cluster_addr=https://172.16.0.1:8201 active_redirect_addr=https://172.16.0.1:8200
Apr 27 16:22:16 master1 vault[22918]: 2018/04/27 16:22:16.160435 [TRACE] core: refreshing forwarding connection
Apr 27 16:22:16 master1 vault[22918]: 2018/04/27 16:22:16.160448 [TRACE] core: clearing forwarding clients
Apr 27 16:22:16 master1 vault[22918]: 2018/04/27 16:22:16.160461 [TRACE] core: done clearing forwarding clients
Apr 27 16:22:16 master1 vault[22918]: 2018/04/27 16:22:16.160508 [TRACE] core: done refreshing forwarding connection
Apr 27 16:22:16 master1 vault[22918]: 2018/04/27 16:22:16.160601 [TRACE] core: creating rpc dialer: host=fw-7def624e-a79c-8092-d6a2-02d95289b43c
Apr 27 16:22:16 master1 vault[22918]: 2018/04/27 16:22:16.160823 [WARN ] grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 172.16.0.1:8201: getsockopt: connection refused"; Reconnecting to {:
172.16.0.1:8201 0  <nil>}
Apr 27 16:22:16 master1 vault[22918]: 2018/04/27 16:22:16.160851 [DEBUG] forwarding: error sending echo request to active node: error=rpc error: code = Unavailable desc = grpc: the connection is unavailable
Apr 27 16:22:17 master1 vault[22918]: 2018/04/27 16:22:17.160742 [TRACE] core: creating rpc dialer: host=fw-7def624e-a79c-8092-d6a2-02d95289b43c
Apr 27 16:22:17 master1 vault[22918]: 2018/04/27 16:22:17.160953 [WARN ] grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 172.16.0.1:8201: getsockopt: connection refused"; Reconnecting to {172.16.0.1:8201 0  <nil>}
Apr 27 16:22:18 master1 vault[22918]: 2018/04/27 16:22:18.562972 [TRACE] core: creating rpc dialer: host=fw-7def624e-a79c-8092-d6a2-02d95289b43c
Apr 27 16:22:18 master1 vault[22918]: 2018/04/27 16:22:18.563191 [WARN ] grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 172.16.0.1:8201: getsockopt: connection refused"; Reconnecting to {172.16.0.1:8201 0  <nil>}
Apr 27 16:22:21 master1 vault[22918]: 2018/04/27 16:22:21.516748 [TRACE] core: creating rpc dialer: host=fw-7def624e-a79c-8092-d6a2-02d95289b43c
Apr 27 16:22:21 master1 vault[22918]: 2018/04/27 16:22:21.516996 [WARN ] grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 172.16.0.1:8201: getsockopt: connection refused"; Reconnecting to {172.16.0.1:8201 0  <nil>}
```

**Steps to Reproduce:**
I belive that the problem occurs when a ZK failure occurs right after the new leader was advertised to the Vault cluster but before leader acquiring process has finished [1].

[1] https://github.com/hashicorp/vault/blob/v0.8.3/vault/core.go#L1583-L1605

**Possible root cause:**

My theory is that the ZK failure was long enough to not to allow subsequent `lock.Unlock()` execute
in the `c.postUnseal()` error handling block but not enough to expire ephemeral nodes made by ZK lock - other Vault instances were not allowed to perform reelection as the leader lock is still there.

Is this something well know? If not - could you please advice - whether the theory/idea for the root cause of this bug is correct and/or whether this is something that could/should be fixed in the `vault/core.go` module itself?

Thanks in advance.

**Comment by jefferai** [comment]
If the error was ephermeral it seems like the zk library or the zk physical implementation should be handling retries there. Additionally the logic in core has changed somewhat in more recent versions of Vault so I don't know if that would make a difference. Vault is ignoring any error coming from the Unlock call, and while I don't konw the genesis of that, my guess is that it's unclear what the right path is if an error is returned.

**Comment by mhrabovcin** [comment]
> Vault is ignoring any error coming from the Unlock call, and while I don't konw the genesis of that, my guess is that it's unclear what the right path is if an error is returned.

This can definitely happen and `vault` is accepting here that lock was never released (so every instance in a cluster cannot get hold of a new lock) and `vault` instance that was a leader is stepping down. This causes a deadlock and only restart of the vault would resolve this situation.

**Comment by jefferai** [comment]
If you have a proposal for addressing this I'm happy to hear it, but I don't think having vault spin on the Unlock call is the right answer. Maybe for a few retries, but if you're trying to shut down Vault, having it hang for potentially long periods of time isn't great.

**Comment by mhrabovcin** [comment]
@jefferai I agree that trying to repeat unlock multiple times wouldn't be acceptable. I'd propose to let each backend handle potential retries. I think that vault should end its operation with `os.Exit()` with non-zero code. Vault is usually launched by `systemd` or some scheduling system and it would probably get restarted. This case can happen only in the HA mode and so there should be other instances ready to step in and resume correct operation.

**Comment by jefferai** [comment]
It's more complicated than that -- we can't simply os.Exit because there are important cleanup operations that can happen during normal shutdown. I think instead it would have to trigger shutdown and pass along an appropriate error.

However, I also don't quite understand how this would help. If the lock is stuck, why does restarting Vault matter, instead of simply trying to create a new one from the host in a running Vault? It sounds like in your scenario the lock is able to be grabbed again by the same host, but then it shouldn't reqiure a restart of Vault to do so.

**Comment by mhrabovcin** [comment]
> However, I also don't quite understand how this would help.

I wasn't clear about that in my comment. This would help in the ZK backend case. I am not familiar with other backends implementations. A ZK lock is implemented by creating a ZK ephemeral node. While a ZK client is connected to ZK, the ZK keeps node around until client disconnects. However there is a SessionTimeout variable that would keep ephemeral node around after short client disconnects.

This is what is happening in a case that @vespian described. The vaults ZK client looses connection to a ZK and tries to remove the lock. Lock removal fails. Vault ignores the error and steps down as a leader. However if underlying ZK client in vault connects back to ZK within the SessionTimeout the node will not get removed.

Here restarting the vault would give a ZK client new session and old one would get expired and the leader election node would get removed.

**Comment by jefferai** [comment]
It feels like the better option is to explicitly close the connection after a step down (perhaps only in the case that the lock fails). Forcing a restart of Vault also means forcing a re-unseal of Vault...it's a very big hammer.

**Comment by mhrabovcin** [comment]
Proposed solution in https://github.com/hashicorp/vault/pull/4569

**Comment by jefferai** [comment]
Closed by #4569
